tests/unit/s2n_keylog_test.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include "testlib/s2n_testlib.h"
+
+#include <unistd.h>
+#include <stdint.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <s2n.h>
+
+#include "crypto/s2n_fips.h"
+
+#include "tls/s2n_connection.h"
+#include "tls/s2n_handshake.h"
+#include "tls/s2n_cipher_preferences.h"
+#include "tls/s2n_cipher_suites.h"
+#include "utils/s2n_safety.h"
+
+static int try_handshake(struct s2n_connection *server_conn, struct s2n_connection *client_conn)
+{
+    s2n_blocked_status server_blocked;
+    s2n_blocked_status client_blocked;
+
+    int tries = 0;
+    do {
+        int client_rc = s2n_negotiate(client_conn, &client_blocked);
+        if (!(client_rc == 0 || (client_blocked && errno == EAGAIN))) {
+            return -1;
+        }
+
+        int server_rc = s2n_negotiate(server_conn, &server_blocked);
+        if (!(server_rc == 0 || (server_blocked && errno == EAGAIN))) {
+            return -1;
+        }
+
+        tries += 1;
+        if (tries == 100) {
+            return -1;
+        }
+    } while (client_blocked || server_blocked);
+
+    uint8_t server_shutdown = 0;
+    uint8_t client_shutdown = 0;
+    do {
+        if (!server_shutdown) {
+            int server_rc = s2n_shutdown(server_conn, &server_blocked);
+            if (server_rc == 0) {
+                server_shutdown = 1;
+            } else if (!(server_blocked && errno == EAGAIN)) {
+                return -1;
+            }
+        }
+
+        if (!client_shutdown) {
+            int client_rc = s2n_shutdown(client_conn, &client_blocked);
+            if (client_rc == 0) {
+                client_shutdown = 1;
+            } else if (!(client_blocked && errno == EAGAIN)) {
+                return -1;
+            }
+        }
+    } while (!server_shutdown || !client_shutdown);
+
+    return 0;
+}
+
+int test_conn(struct s2n_config *server_config, struct s2n_config *client_config) {
+    struct s2n_connection *client_conn;
+    struct s2n_connection *server_conn;
+    server_conn = s2n_connection_new(S2N_SERVER);
+    notnull_check(server_conn);
+    int server_to_client[2];
+    int client_to_server[2];
+
+    /* Create nonblocking pipes */
+    GUARD(pipe(server_to_client));
```

Comment:
This will not set s2n_errno, and will confuse whoever is trying to debug test failures.